doc/neps/nep-0055-string_dtype.rst

```diff
@@ -394,133 +399,271 @@ existing concept of an array of references in NumPy beyond the ``object`` DType
+allocator mutex before any usage of the ``allocator``. This prevents parallel
+access to the heap memory used by string arrays.
+
+The ``PyThread_type_lock`` mutex is relatively heavyweight and does not provide
+more sophisticated locking primitives that allow multiple simultaneous
+readers. As part of the GIL-removal project, CPython is adding new
+synchronization primitives to the C API for projects like NumPy to make use
+of. When this happens, we can update the locking strategy to allow multiple
+simultaneous reading threads, along with other fixes for threading bugs in NumPy
+that will be needed once the GIL is removed.
+
+If a mutated string remains the same size or is shrunk, the existing short
+string or arena allocation is re-used, with padding zeros written to the end of
+the subset of the buffer reserved for the string. If the string is enlarged, the
+existing space in the arena buffer cannot be used, so instead we resort to
+allocating space directly on the heap via ``malloc`` and the
+``NPY_STRING_ON_HEAP`` flag is set. Any pre-existing flags are kept set to allow
+future use of the string to determine if there is space in the arena buffer
+allocated for the string for possible re-use.
+
+Freeing Strings
++++++++++++++++
+
+Existing strings must be freed before discarding or re-using a packed
+string. The API is constructed to require this for all strings, even for short
+strings with no heap allocations. In all cases, all data in the packed string
+are zeroed out, except for the flags, which are preserved except as noted below.
+
+For strings with data living in the arena allocation, the data for the string in
+the arena buffer are zeroed out and the ``NPY_STRING_ARENA_FREED`` flag is set
+on the packed string to indicate there is space in the arena for a later re-use
+of the packed string. Heap strings have their heap allocation freed and the
+``NPY_STRING_ON_HEAP`` flag removed.
+
+Memory Layout Examples
+++++++++++++++++++++++
+
+We have created illustrative diagrams for the three possible string memory
+layouts. All diagrams assume a 64 bit little endian architecture.
+
+.. image:: _static/nep-0055-short-string-memory-layout.svg
+
+Short strings store string data directly in the array buffer. On little-endian
+architectures, the string data appear first, followed by a single byte that
+allows space for four flags and stores the size of the string as an
+unsigned integer in the final 4 bits. In this example, the string contents are
+"Hello world", with a size of 11. The only flag set indicates that this is a
+short string.
+
+.. image:: _static/nep-0055-arena-string-memory-layout.svg
+
+Arena strings store string data in a heap-allocated arena buffer that is managed
+by the ``StringDType`` instance attached to the array. In this example, the
+string contents are "Numpy is a very cool library", stored at offset ``0x94C``
+in the arena allocation. Note that the ``size`` is stored twice, once in the
```

Comment:
This is stored to re-use the arena allocation later. We need to store the size to see if a future allocation for the string that owns space in the arena will fit. This only happens if the string is mutated to be the same size or smaller than the original contents.

We could simplify the design by not storing the sizes and not reusing space in the arena when we can and instead whenever a string is mutated to be larger than the original contents always do a heap allocation and never try to reuse the space in the arena after that. That might or might not save memory depending on the situation by not storing the sizes in the arena buffer.

Reply:
Yeah, I figured it was something like this. It perhaps is a bit more detail than needed, since in the end there can be multiple allocators. Since you wrote the text and it explains things nicely, do keep it, but perhaps follow @seberg's suggestion for a bit of a reorganization.

Reply:
Yeah I'll reorganize and try to make sure the points you got tripped up on clearer. Thanks for the read!

Reply:
I added a paragraph explaining why we store sizes in the arena and pointing out that we could also make a simpler choice if we'd like at the cost of doing heap allocations more often when a string is mutated